USWDS - Accordions: Add support for hidden until found

mejiaj opened the pull request:
> [!IMPORTANT]  
> Breaking change for all accordions in IE11. Due to `:where(:not())` selector.

# Summary

**Accordion content is now discoverable.** As an enhancement, content inside accordions is now discoverable. Browsers must support `hidden="until-found"` attribute[^1]. 

Browsers that **don't** support this attribute will fallback to current `hidden` behavior.

[^1]: [Caniuse.com](https://caniuse.com/mdn-html_global_attributes_hidden_until-found_value)


## Breaking change

This is potentially a breaking change. Users with padding overrides on accordions are affected.

**Broken when custom padding set**
USWDS website examples have padding overrides.

| Before | After |
| :----- | :---- |
| ![image](https://github.com/uswds/uswds/assets/3385219/231bf1c3-4209-4610-9526-775964d8b92e) |  ![image](https://github.com/uswds/uswds/assets/3385219/68c4b3f2-ce92-4e3a-93b2-f9f8e60ca1ce)  | 

**Browsers tested**

| Browser         | Tested version | Latest version |
| --------------- | -------------- | -------------- |
| Chrome          | 112            | 119            |
| Firefox         | 112            | 120            |
| Opera           | 97            |      105       |
| Edge            | 112            | 119            |
| iPhone (Safari) | 12             | 15             |

_Based on https://analytics.usa.gov/ data._

## Related issue

Closes #4799.
Closes uswds/uswds-site#2405.

## Related pull requests

TODO: Need changelog PR.

## Preview link

- [Accordion test preview →](https://federalist-3b6ba08e-0df4-44c9-ac73-6fc193b0e19c.sites.pages.cloud.gov/preview/uswds/uswds/jm-feature-accordion-hidden-content/?path=/story/components-accordion--test-icons) 
- [Header preview →](https://federalist-3b6ba08e-0df4-44c9-ac73-6fc193b0e19c.sites.pages.cloud.gov/preview/uswds/uswds/jm-feature-accordion-hidden-content/?path=/story/components-header--default)
- [Secondary nav preview →](https://federalist-3b6ba08e-0df4-44c9-ac73-6fc193b0e19c.sites.pages.cloud.gov/preview/uswds/uswds/jm-feature-accordion-hidden-content/?path=/story/components-header-partials-secondary--default)


## Problem statement

Accordion content should be discoverable on page search. Currently, content is hidden. Users find it difficult to find content in of accordions. There's also an impact to SEO.

## Solution

Enhanced toggle utility (used in Accordion, Banner, Header, and Language Selector[^2]). This approach is a progressive enhancement on existing accordions. Unsupported browsers like Safari and Firefox will fallback to previous behavior.

[^2]: Language selector is unaffected because it uses `hidden="true"`.

## Major changes

- Remove padding on hidden content  with`:where(:not())` selector.
- Add check to toggle.js for hidden until found, fallback to hidden.
- Normalize supports hidden until found.
- Accordion test includes header variants.
- Header menus include secondaryNav data in JSON (previously as StorybookJS args) fixes uswds/uswds-site#2403.
- Header hides secondary nav & search if not defined.

## Testing and review

**Chrome**

1. Visit [Preview →](https://federalist-3b6ba08e-0df4-44c9-ac73-6fc193b0e19c.sites.pages.cloud.gov/preview/uswds/uswds/jm-feature-accordion-hidden-content/?path=/story/components-accordion--test-icons)
2. Banner
    - Search for a term in content (like "_government organization_").
    - Component should expand
3. Accordion
    - Search for a term in content (like "_otherwise infamous_").
    - Fifth amendment should expand


**Firefox or Safari**
These browsers don't support `hidden="until-found"`. There should be no regressions in behavior or display.

**Header changes**
1. Run `npm run build:html`.
2. View compiled HTML file in `html-templates` and confirm secondary nav is showing.
3. Visit StorybookJS preview and ensure there are no regressions in:
    - [Header](https://federalist-3b6ba08e-0df4-44c9-ac73-6fc193b0e19c.sites.pages.cloud.gov/preview/uswds/uswds/jm-feature-accordion-hidden-content/?path=/story/components-header--default)
    - [Header megamenu](https://federalist-3b6ba08e-0df4-44c9-ac73-6fc193b0e19c.sites.pages.cloud.gov/preview/uswds/uswds/jm-feature-accordion-hidden-content/?path=/story/components-header--megamenu)
    - [Secondary nav](https://federalist-3b6ba08e-0df4-44c9-ac73-6fc193b0e19c.sites.pages.cloud.gov/preview/uswds/uswds/jm-feature-accordion-hidden-content/?path=/story/components-header-partials-secondary--default)

- [x] Confirm that this code follows the [18F Front End Coding Style Guide](https://pages.18f.gov/frontend/) and [Accessibility Guide](https://pages.18f.gov/accessibility/checklist/).
- [x] Run `npm run prettier:sass` to format any Sass updates.

<!--
Before opening this PR, make sure you’ve done whichever of these applies to you:
- [ ] Run `git pull origin [base branch]` to pull in the most recent updates from your base and check for merge conflicts. (Often, the base branch is `develop`).

- [ ] Run `npm test` and confirm that all tests pass.
- [ ] Run your code through [HTML_CodeSniffer](http://squizlabs.github.io/HTML_CodeSniffer/) and make sure it’s error free.
-->

Commit:
Add support for hidden until found; closes #4799

- Remove padding on hidden content  with`:where(:not())` selector.
- Add check to toggle.js for hidden until found, fallback to hidden.
- Normalize supports hidden until found.

mahoneycm reviewed (commented):
This is awesome! Really loving the functionality. I hope more browsers add support for this sooner than later.

I had an idea for a potential enhancement and a css question, but otherwise lgtm

### Testing Checklist

- [x]  No visual regression
- [x]  Using `beforematch` is a logical test for hidden until found support
- [x]  In page search properly locates previously hidden content
- [x]  Unsupported browsers behave as expected

mejiaj commented on packages/usa-banner/src/styles/_usa-banner.scss:
```diff
@@ -80,18 +80,21 @@ $banner-icon-close: (
 
-  @include at-media("tablet") {
-    @include u-padding-y(3);
+  &:where(:not([hidden])) {
```
Yes, it's a pattern we've previously agreed to because it could introduce unwanted specificity issues.

amyleadem commented on packages/uswds-elements/lib/_normalize.scss:
```diff
@@ -349,3 +349,8 @@ template {
+
+[hidden="until-found"] {
+  display: revert;
+  content-visibility: hidden;
```
Curious - does this rule need to be explicitly defined? It looks like the browser implicitly applies it with `hidden="until-found"` and removing it doesn't seem to have an affect.  

> hidden=until-found applies the content-visibility:hidden CSS property instead of the display:none property that the regular hidden attribute applies. [^1]

[^1]: [developer.chrome.com](https://developer.chrome.com/docs/css-ui/hidden-until-found#caveats)

```suggestion
```

mejiaj replied:
@amyleadem this was added explicitly for clarity. Especially as this is a new feature. Please let me know if you'd like me to remove.

amyleadem reviewed (changes requested):
This is a really nice enhancement for browsers that support it. I think the way you built it is smart. Also, thanks for the excellent PR description and footnotes! 

I found one visual regression after pulling in the updates from `develop`. I've added a screenshot below. I also had a couple questions about the Sass added to `normalize.scss`, but I don't consider either question to be a blocker.

Here are the tests I performed:

- [x] In Chrome, confirm that page search finds accordion content and opens the related accordion panel
    - In the preview link, I tested a search for the word “shall”. It found 9 instances and opened each panel as I moved through the results. In `develop`, it finds only one instance of the word.
    - Also tested the following components: accordion, banner, header, and language selector
- [x] Confirm that Chrome components receive `hidden="until-found` and other browsers receive `hidden`
- [ ] Confirm no visual regressions in any browser
    - Everything looks good in Firefox and Safari.
    - In Chrome, I noticed a regression in the header component after pulling  `develop` into this PR
    ![image](https://github.com/uswds/uswds/assets/93996430/91363737-b717-4c30-8180-9af38c8725eb)
- [x] Confirm that VoiceOver still works as expected in all browsers

amyleadem commented:
Also, I noticed that this isn't tagged for a release. What do you think about adding it to 3.8.0? (I know we don't need more in there (🙃), but I just really like the feature).

mejiaj commented on packages/uswds-elements/lib/_normalize.scss:
```diff
@@ -349,3 +349,8 @@ template {
   display: none;
 }
+
+[hidden="until-found"] {
```
Browser support for `until-found` didn't pick-up until 2022 via _Can I use_.

<details><summary>Support screenshot</summary>
<p></p>

<img width="1000" alt="image" src="https://github.com/uswds/uswds/assets/3385219/f175bf15-6f15-47c1-8241-ad4255aebaa3">

**Source:** https://caniuse.com/mdn-html_global_attributes_hidden_until-found_value (_Hit the `Date relative` button_)

</details> 


The last normalize release was Nov 4, 2018, via [Release 8.0.1 · necolas/normalize.css](https://github.com/necolas/normalize.css/releases/tag/8.0.1). 

**Related issues**
- #5555
- #4701

mejiaj commented:
> Also, I noticed that this isn't tagged for a release. What do you think about adding it to 3.8.0? (I know we don't need more in there (🙃), but I just really like the feature).

Yes, I'd like to see it too. I originally added it to the November milestone to not increase the size of 3.8.0.

Should we add it to 3.8.0 @thisisdano?

Commit:
Accordion: improve testing and fix visual regression

- Fix visual regression where megamenu padding would make the display visible when it hidden
- Create a story for testing all instances of accordion.

Commit:
Fix secondary nav HTML build in header variants. Fixes uswds/uswds-site#2403.

- SecondaryNav data was moved to the header JSON file to fix broken HTML showing in code examples

Commit:
Remove unused JSON file

Commit:
Remove extra variable in search

Commit:
Use longform escape for clarity

Commit:
Header: check before displaying search and secondary nav

Commit:
Update accordion test template

Removing menus and editing aria labels to avoid a11y errors in builds

Commit:
A11Y adjustments to test for CI

mejiaj commented on packages/usa-header/src/_includes/usa-navbar.twig:
```diff
@@ -1,7 +1,5 @@
-  {% include "@components/usa-site-title/src/usa-site-title.twig" with {
-    'site_title': '<Project title>'
-    }
+  {% include "@components/usa-site-title/src/usa-site-title.twig" with site_title
```
Removed this because we were using `default()` in `usa-site-title` anyways.

mejiaj commented on packages/usa-header/src/usa-header--extended/usa-header--extended-megamenu.json:
Resolves display issue in website.

mejiaj commented on packages/usa-header/src/usa-header--extended/usa-header--extended.json:
Resolves display issue in website.

mejiaj commented on packages/usa-header/src/usa-header--extended/usa-header--extended.twig:
Avoids displaying a broken secondary nav if menu is disabled.

mejiaj commented on packages/usa-header/src/usa-header.twig:
```diff
@@ -5,7 +5,9 @@
-      {% include "@components/usa-search/src/usa-search.twig" with search.settings %}
+      {% if search.settings %}
+        {% include "@components/usa-search/src/usa-search.twig" with search.settings %}
+      {% endif %}
```
For testing purposes, gives developers option of excluding search.

mejiaj commented on packages/usa-search/src/usa-search.twig:
Avoids unnecessary variable.

mejiaj commented on packages/usa-accordion/src/usa-accordion.stories.js:
```diff
@@ -21,9 +28,47 @@ Bordered.args = BorderedContent;
+    },
+    search: null,
+    secondaryNav: false,
+  },
```
A lot of overrides here to have a simple menu. Header needs refactoring to get it in a more customizable state.

Commit:
Site title: set default before escaping

mahoneycm reviewed (approved):
Looking good across browsers at different breakpoints!

- [x]  No visual regression
- [x]  Using `beforematch` is a logical test for hidden until found support
- [x]  In page search properly locates previously hidden content
- [x]  Unsupported browsers behave as expected

[^1]: [Additional info in comment](https://github.com/uswds/uswds/pull/5644/files#r1410801278)

>[!NOTE]
> Hoping to listen to `onbeforematch` to toggle accordions appropriately as a future enhancement [^1]
